main.go

```diff
@@ -13,122 +13,143 @@
+	flagMaxTime  = flags.Duration("maxtime", 0, "maximum time to run")
+	flagMaxRuns  = flags.Int("maxruns", 0, "maximum number of runs")
+	flagMaxFails = flags.Int("maxfails", 1, "maximum number of failures")
+	flagStdErr   = flags.Bool("stderr", true, "output failures to STDERR instead of to a temp file")
 )
 
 func roundToSeconds(d time.Duration) time.Duration {
 	return time.Duration(d.Seconds()+0.5) * time.Second
 }
 
-func main() {
-	flag.Parse()
-	if *flagP <= 0 || *flagTimeout < 0 || len(flag.Args()) == 0 {
-		flag.Usage()
-		os.Exit(1)
+func run() error {
+	if err := flags.Parse(os.Args[1:]); err != nil {
+		return err
+	}
+	if *flagP <= 0 || *flagTimeout <= 0 || len(flags.Args()) == 0 {
+		var b bytes.Buffer
+		flags.SetOutput(&b)
+		flags.Usage()
+		return errors.New(b.String())
 	}
 	var failureRe, ignoreRe *regexp.Regexp
 	if *flagFailure != "" {
 		var err error
 		if failureRe, err = regexp.Compile(*flagFailure); err != nil {
-			fmt.Println("bad failure regexp:", err)
-			os.Exit(1)
+			return fmt.Errorf("bad failure regexp: %s", err)
 		}
 	}
 	if *flagIgnore != "" {
 		var err error
 		if ignoreRe, err = regexp.Compile(*flagIgnore); err != nil {
-			fmt.Println("bad ignore regexp:", err)
-			os.Exit(1)
+			return fmt.Errorf("bad ignore regexp: %s", err)
 		}
 	}
+
+	c := make(chan os.Signal)
+	defer close(c)
+	signal.Notify(c, os.Interrupt)
+	defer signal.Stop(c)
+	var wg sync.WaitGroup
+	defer wg.Wait()
+	ctx, cancel := func() (context.Context, context.CancelFunc) {
+		ctx := context.Background()
+		if *flagMaxTime > 0 {
+			return context.WithTimeout(ctx, *flagMaxTime)
+		}
+		return context.WithCancel(ctx)
+	}()
+	defer cancel()
+	go func() {
+		for range c {
+			cancel()
+		}
+	}()
+
 	startTime := time.Now()
-	finishTime := time.Now().Add(*flagMaxTime)
+
 	res := make(chan []byte)
-	var exitFlag int32
+	wg.Add(*flagP)
 	for i := 0; i < *flagP; i++ {
-		go func() {
-			for atomic.LoadInt32(&exitFlag) == 0 {
-				cmd := exec.Command(flag.Args()[0], flag.Args()[1:]...)
-				done := make(chan bool)
-				if *flagTimeout > 0 {
-					go func() {
-						select {
-						case <-done:
-							return
-						case <-time.After(*flagTimeout):
-						}
-						if !*flagKill {
-							fmt.Printf("process %v timed out\n", cmd.Process.Pid)
-							return
-						}
-						cmd.Process.Signal(syscall.SIGABRT)
-						select {
-						case <-done:
-							return
-						case <-time.After(10 * time.Second):
+		go func(ctx context.Context) {
+			defer wg.Done()
+			for ctx.Err() == nil {
+				res <- func() []byte {
+					var cmd *exec.Cmd
+					if *flagTimeout > 0 {
+						if *flagKill {
+							var cancel context.CancelFunc
+							ctx, cancel = context.WithTimeout(ctx, *flagTimeout)
+							defer cancel()
+						} else {
+							defer time.AfterFunc(*flagTimeout, func() {
+								fmt.Printf("process %v timed out\n", cmd.Process.Pid)
+							}).Stop()
 						}
-						cmd.Process.Kill()
```

Comment:
This and the `cmd.Process.Signal(syscall.SIGABRT)` seem important. How else do you ensure rogue test processes that hang forever get killed?